### ISSUE TITLE
Support PIT - the Java mutation testing framework

### ISSUE THREAD

**Issue by nicojs** [stryker-mutator/mutation-testing-elements#18]
[PIT](https://github.com/hcoles/pitest) is interested in using mutation testing elements  see https://twitter.com/0hjc/status/1113369365066473473

We need to support mutants that don't have an exact location or replacement. Only line number and mutator name.

Example JSON:

```json
{ 
  "schemaVersion": "1.1",
   "thresholds": {
    "high": 80,
    "low": 60
  },
    "files": {
       "src/main/foo.java": {
           "language": "java",
           "source": "public class Progam {\n    public static void main(String[] args) {\n    System.out.println(\"bar\");\n}",
          "mutants": [
            {
              "id": "1",
              "mutatorName": "StringLiteral",
              "location": {
                "start": {
                  "line": 3
                }
              },
              "status": "Survived"
            }
          ]
       }
```

@hcoles 
Is this json something PIT could produce? Also: mutation-testing-elements is opinionated: `"mutatorName"` is a free text field, but `"status"` isn't. Do you agree with our definitions and metrics? Documented here: https://github.com/stryker-mutator/stryker-handbook/blob/master/mutant-states-and-metrics.md

**Comment by hugo-vrijswijk** [comment]
Requiring a replacement when a column is present, but not otherwise might be able to be enforced in the schema by using [property dependencies](https://json-schema.org/understanding-json-schema/reference/object.html#dependencies)

**Comment by hugo-vrijswijk** [comment]
This might also be a good time to look at a 'real' release to maven central, instead of using [webjars](http://webjars.org/)

**Comment by nicojs** [comment]
I would separate the two issues. Feel free to add an automated release to maven 💪.

**Comment by hcoles** [comment]
@nicojs 

The status definitions match very closely to what PIT does (even the names match). We have the additional (rare) status of "memory error" but this can be mapped to runtime error.

As PIT can't provide a replacement it provides a description of the mutation in addition to the mutator name. For some mutators this can provide useful contextual information, for others it is a fixed description.

The mutator name and description could perhaps be concatenated together to fit this schema, but from my point of view it would be preferable to keep them separate if possible.

**Comment by hcoles** [comment]
I've created an issue on the pitest side. 

https://github.com/hcoles/pitest/issues/592

Unfortunately it will need to wait on finding some-one with more time than I have to look at it.

**Comment by nicojs** [comment]
@hcoles thanks for the input. Indeed, adding a description would be preferred in this case. We could show it in de the popup when clicking on the mutant. I will create a PR for it. We could release it in a v1.1 of the schema, making location `column`s and `end` optional as well.

Glad to hear the names of the mutant states are in sync, that makes it easier for both integrating today as well as using it as an end user 👍 

As for the amount of time you have, maybe one of the Stryker4s team members want to help? @hugo-vrijswijk @legopiraat @Wmaarts

**Comment by Wmaarts** [comment]
https://github.com/Wmaarts/pitest-stryker-report-plugin. 🎉

**Comment by Wmaarts** [comment]
There's still a couple of things lingering related to this.

First, as @nicojs said:
> We need to support mutants that don't have an exact location or replacement. Only line number and mutator name.

Correct, since Pitest doesn't provide a column it'd be nice to be able to just provide the line number. Currently there is a workaround that sets `start.column` to 1 and `end.column` to the length of the string that is that line of code. [here](https://github.com/Wmaarts/pitest-stryker-report-plugin/blob/742f1daafcb5dfe59d90575912360dccad84f2a0/src/main/java/org/pitest/stryker/models/json/StrykerLocation.java#L19)

Second, again, as @nicojs said:
> Indeed, adding a description would be preferred in this case. We could show it in de the popup when clicking on the mutant.

Currently the field 'replacement' is filled with the description given by Pitest. A description field would be much nicer. :)

Third, a question to @hcoles:
While running Pitest, I ran into the issue that I received `MutationResult`s for certain files, but I could not retrieve the source code of these files. =/ Is this regular behaviour?

(Pitest HTML reporter has no issue with this since it lists all mutants in a file at the bottom of anyway, but this HTML reporter does not do this, so we'd have to think of something.)

**Comment by hcoles** [comment]
@Wmaarts Can you clarify what you mean by "not retrieve the source code"? Is there no filename present or is there a filename that doesn't resolve to a source file?

**Comment by hcoles** [comment]
Having had a look at the output I've realised something important is missing.

Pitest displays mutation coverage and normal line coverage side by side. This makes it very easy to scan a file a find interesting mutants (surviving ones with coverage).

I suppose the same information is captured by the "no coverage" status, but this would need to be displayed prominently without the need to click on the mutant if it were to be used to quickly examine a report.

**Comment by Wmaarts** [comment]
> Can you clarify what you mean by "not retrieve the source code"?

I think it's specifically when this method returns an empty Optional -> [here](https://github.com/hcoles/pitest/blob/57af81af9e4914c4f89c34af485c2342ec6e6a9e/pitest-html-report/src/main/java/org/pitest/mutationtest/report/html/MutationHtmlReportListener.java#L164).

So I think when a filename doesn't resolve to a source file.

**Comment by hcoles** [comment]
@Wmaarts This shouldn't happen unless the some sort of code generator is being used (that said I have vague memories of some corner cases around lambdas).

Can you provide an example of where you're seeing this?

**Comment by Wmaarts** [comment]
It's probably my mistake then. It only happened when I ran pitest with the command-line entrypoint. I probably used incorrect parameters?

**Comment by hcoles** [comment]
@Wmaarts That would make sense, if the source directories aren't properly configured pitest won't be able to find the files. It's pretty easy to mess the config up if you're wiring things by hand.

**Comment by nicojs** [comment]
@hcoles 

> Pitest displays mutation coverage and normal line coverage side by side. This makes it very easy to scan a file a find interesting mutants (surviving ones with coverage).

You mean this right?

![image](https://user-images.githubusercontent.com/1828233/57284555-b2a21600-70b1-11e9-9f03-5910ccbdd2ba.png)

Question: is this line coverage calculated as actual code coverage? Or could it be calculated using existing metrics from https://github.com/stryker-mutator/stryker-handbook/blob/master/mutant-states-and-metrics.md

For example, we could use `Covered / Valid * 100`? Or is this more something like `mutantCoverage`? If this is not the same as PIT's line coverage, would it be a valid equivalent? Or do we really want to include actual code coverage in our JSON format?

**Comment by Wmaarts** [comment]
> Currently there is a workaround that sets start.column to 1 and end.column to the length of the string that is that line of code.

I'm actually thinking this workaround isn't a big deal, we won't need to update the schema to support just lines without columns and we can just keep it this way.

**Comment by hcoles** [comment]
Unfortunately it can't be calculated from the mutation statistics - a line with coverage may not have a mutant seeded.

The thing that makes the line coverage useful is not the high % metric, it's the qualitative information it gives you when looking at an annotated source file - it makes it easy to quickly spot interesting mutants (i.e. mutants on lines with coverage).

Most of this benefit could be achieved by making mutants without coverage clearly visibly distinct from ones that survived with coverage. At the moment you need to click in order to distinguish them.

**Comment by nicojs** [comment]
Ok, I think I understand. You mean this part:

![image](https://user-images.githubusercontent.com/1828233/57296445-51883b80-70cd-11e9-8fbd-5a5cea421335.png)

(I just now realized that the coloring of the line number represented the line coverage)

You can also see this per mutant at a glance using the filters on top: 

![image](https://user-images.githubusercontent.com/1828233/57296619-afb51e80-70cd-11e9-9a18-fac9cf0e8e44.png)

In this example, the `help` property wasn't covered. The mutants aren't clickable here, because the `NoCoverage` status is not checked in the filters on top.

We could also introduce a new color for status "NoCoverage", but I don't think it's strictly necessary. The information is there for anyone interested.

**Comment by nicojs** [comment]
@hcoles any updates on this? If we know what to do, we can improve the report, but as stated in my previous remark, I don't think it's needed.

**Comment by hcoles** [comment]
@nicojs If the page opens by default showing **only** the surviving mutants then I think this gives most of the benefit - you just need to open a page to see the interesting mutants without having to click anything.

Having a color coding for survived vs uncovered vs killed might also be nice (not looking at the report just now so can't remember if this is already the case).

**Comment by nicojs** [comment]
> If the page opens by default showing **only** the surviving mutants then I think this gives most of the benefit - you just need to open a page to see the interesting mutants without having to click anything.

Right now we show `survived` and `no coverage` by default when you open a report. For example:

![image](https://user-images.githubusercontent.com/1828233/58692464-edd0f580-838e-11e9-892e-f129e357bc93.png)

I think this is fine, both deserve attention, right? However, distinguishing them with color seems like a good idea. The same example with different in colors (`no coverage` is yellow, the same as `timeout` or `error`):

![image](https://user-images.githubusercontent.com/1828233/58692776-a1d28080-838f-11e9-9c55-fd2c885a1f46.png)

Let's make the following changes:

* Choose a color for `no coverage` (not the same as `timeout` or `error`).
* Also show the colors in the file legend above (instead of only the name). Probably show it in a badge.

@hcoles do you agree with this?

**Comment by hcoles** [comment]
@nicojs Sounds good. Perhaps different shapes for the icon on the left as well?

**Comment by nicojs** [comment]
What do you mean exactly? It is now shown as a badge with a number in it. You want to change that? @Wmaarts introduced an emoj per status in the popup:

* MutantStatus.Killed: '✔';
* MutantStatus.NoCoverage: '❌';
* MutantStatus.Survived:
* MutantStatus.Timeout: '⌛';
* MutantStatus.RuntimeError:  '⚠';
* MutantStatus.CompileError: '⚠';

Do you want to use that inside the badge?

**Comment by nicojs** [comment]
Edit: @Wmaarts will update the pitest mutation testing elements plugin so that it will add the badge to the first non-whitespace character of the line. So it will be more in line with other implementations.

**Comment by hcoles** [comment]
@nicojs I mean some visual indication of the status apart from color (good for the color blind and folk with other visual issues).

I was imagining something subtle like square badges with a number for survivors, triangle for no coverage, circles for kills etc, but really just throwing out the idea of using something in addition to color to make the difference stand out.

**Comment by nicojs** [comment]
> I was imagining something subtle like square badges with a number for survivors, triangle for no coverage, circles for kills etc

Why use geometric figures when you can use emojies?

How about this?
![image](https://user-images.githubusercontent.com/1828233/58704874-bf164780-83ad-11e9-9c0d-c5bb744d156d.png)

And opening a mutant:

![image](https://user-images.githubusercontent.com/1828233/58704955-01d81f80-83ae-11e9-989c-a83dec04b857.png)

**Comment by nicojs** [comment]
Proposed list of emojies:

* Killed: '✅'
* NoCoverage: '🔬'
* Survived: '👽'
* Timeout: '⌛'
* RuntimeError: '💥'
* CompileError: '💥'

We can also go for more traditional:

* Killed: '✅'
* NoCoverage: '⚠'
* Survived: '❌'
* Timeout: '🕓'
* RuntimeError: '⛔'
* CompileError: '⛔'

Or a combination of the two. 🤷‍♂

**Comment by nicojs** [comment]
Alternative for NoCoverage: 👓. It really is the most tricky to find an emoji for.

**Comment by mthmulders** [comment]
How about 🙈  for _no coverage_?

**Comment by hugo-vrijswijk** [comment]
Having emojis per category at the top is nice, I think. Having them on each line feels a bit like emoji vomit 🤮

**Comment by nicojs** [comment]
> Having them on each line feels a bit like emoji vomit 🤮

Yeah, that's a good point. However, @hcoles 's point about the visually impaired is also valid... We could also make it optional (which adds to the complexity)

**Comment by hugo-vrijswijk** [comment]
@hcoles Last week a new version of the [PIT mutation-testing-elements reporter](https://github.com/Wmaarts/pitest-mutation-testing-elements-plugin) was released. Feel free to have a look and discuss how you'd like to proceed. I've also uploaded the new report (ran on the plugin of course) here:

https://send.firefox.com/download/66d35da876f814f1/#bpZTp8WUvXG1smHrdbEeGg

**Comment by Mobrockers** [comment]
@hugo-vrijswijk @nicojs @Wmaarts @hcoles  Since there is a plugin, can this issue be closed? If additional work is required for better compatibility of the schema with pitest I think we should open a new issue with a list of needed changes.

**Comment by hugo-vrijswijk** [comment]
Yeah I would say so
